### PR TITLE
Improved Inspector.

### PR DESCRIPTION
**Pull request by djrm**
Fixed a lot of metrics issues
Added a better look for subresource explorer (colors need a bit of work)
Fixed some theme initialization problems

this need some minor improvements that i plan to address in other PRs.
|peek 2018-05-22 16-25 | peek 2018-05-22 16-26|

### CODE REVIEW

**Comment by groud** [comment]
Nice it looks awesome !

To the list of the things that I still find a little bit weird I see:
- Sliders not taking the full available horizontal room
- The color properties boxes are not the same size as the other widgets
- There's something odd with the vector boxes. I feel like a margin between the two values is missing. Or maybe the fact they are misaligned with the other widget makes it feel wrong I don't know. (I wonder how it is done elsewhere).
- The colors for the subresource editor do not make it feel like a "block". Could you try by changing the background color instead of the text color for the edited subresource header ? Also, I'm curious to see if it would look better by not discontinuing the line when opening a subresource.

Still it's a very good job ! :D

**Comment by djrm** [comment]
@groud i tried doing the background change, but it's not good, also color and other detail will be fixed with the theme update.

**Comment by cart** [comment]
Love it!  Have you considered using the text input style for the checkboxes? Using the same style helps people think "this is something I can click to edit the state".

This shows both my proposal (Flip H) and the current state (Flip V) 
checkbox

**Comment by toger5** [comment]
some things are amazing!! I need to try it out tomorrow but I think it already is much easier to have an overview of whats going on.

opinions:
 - there are too many fonts. bold small large saturated., grayed out. Of course some diversity is needed but Imo it's too much. and look a little chaotic
 - The editor looks like this one vertical worm. maybe there is a way to give it more horizontal sepertion items (somehow I also don't know what would work best) but right now I find it hard to distinguish section headers from values or sub recourse label.

I like this example since it succeeds in making clear horizontal separations between sections. while still having a high input_field density.
(But it is much less complex than the prop hierarchy the godot editor has to deal with...)
https://github.com/Alecaddd/Akira

**Comment by djrm** [comment]
@toger5 yeah that akira thing looks really nice, but yeah godot is way more complex and its UI toolkit not as flexible, actually modifying the inspector was a pain, but still i want to improve the theme, and that program  gave me a some ideas, will share some mockups soon.

**Comment by sersoong** [comment]
The new features are great.
But it looks a bit ugly.
Such as vector2's x and y color,they can be the same color.
And the background color of the class title ,it looks so ugly.Blinding my titanium dog eye.

**Comment by Geequlim** [comment]
The indent lines are really useful to recongnize sub-resources. It would be better if the color of the lines could be less bright. Here is the indent lines of VSCode.
image

In my opinon the background colors of the inspector in 3.0 is good to recongnize the level of classes, groups and properties form brighter to darker. We don't should to change this user expirence.

image

I'm agree with @sersoong that we don't need to put too much colors into the inspector for the elements of vectors.

**Comment by damarindra** [comment]
agree with @sersoong , please don't change x and y color

**Comment by groud** [comment]
I personnaly like the new X and Y colors corresponding to the axis colors. That's a great improvement.

Also I kind of agree with @Geequlim. I think it might be better to keep the dark background for properties, and lighter for the headers. I think that the fact that there's a strong difference between the LineEdit color and the background makes things less easy to understand as a "block".

> @groud i tried doing the background change, but it's not good, also color and other detail will be fixed with the theme update.
 
Do you have a screenshot maybe ? Else I thiink we need something stronger than the text color to understand that the header is a header of the colored block.

**Comment by groud** [comment]
Oh also, I saw that you aligned the class headers to the left. I think they are better in the center that makes a better difference between inner categories.

**Comment by neikeq** [comment]
> Oh also, I saw that you aligned the class headers to the left. I think they are better in the center that makes a better difference between inner categories.

That. There was something in these screenshots that made the inspector structure very hardto follow, but I couldn't find what it was. The class header aligned to the center definitely improves things.

**Comment by DriNeo** [comment]
There is so many settings in the inspector that is challenging to make it very readable. Many thanks to djrm for working on this !
A solution is to pop a kind of flying dock for every property (like Gimp panels as instance ?). But maybe I'm wrong of course.

**Comment by mhilbrunner** [comment]
Looks awesome. I really like that the X/Y keys have the axis colors. Will saves new users quite some time.

**Comment by eon-s** [comment]
The problem with colors in the inspector are the different themes, it is possible to change a font theme but will need special settings to change these particular vector colors.

**Comment by djrm** [comment]
Ok, thanks for the feedback, as i said i plan to fix all the minor issues in the upcoming theme update (most of them are already considering in my mockups), so i would like to have this merged so people can actually test it, and give proper feedback. maybe here #19084

**Comment by groud** [comment]
Honestly since it's mostly a visual improvement I think it's better if you include at least the consensual feedback in this PR. I think it is not a problem if the PR gets a little bit big.

**Comment by djrm** [comment]
@groud no, the problems are actually possible conflicts and overlapping work between this and upcoming theme changes, i mean this does not go directly to production, so i dont see problems with improving when doing the theme changes, also i see most of the feedback comes from watching the GIFs which are not that good and quite confusing would be better to have actual usage feedback, on the other hand there is no feedback consensus.

**Comment by groud** [comment]
Well... since it's a mostly visual update I'm not sure usage feedback is that important. It's not common for us to merge something that is a WIP unless it is an improvement (which it not obvious for now, it looks less readable than before).

But we'll see what others think on the PR meeting I guess :)

**Comment by NathanLovato** [comment]
The XY(Z) colors are a good idea. I also prefer to have them side-by-side. It reuses the colors from the viewport, makes it easy to spot these essential properties for animators and designers.

The class names aligned left doesn't read as well as the current centered ones. In Godot 3 I know what is interactive (text aligned left) and what's a section header (centered) at a glance:

image

Regarding the relationship lines, agreed it needs testing. @groud that's a UX change, it'll affect the way people interact with the inspector even if it's just visual. Looking at it I feel Juan's initial design with larger margin and a border-left reads a lot better for sub-resources - they do feel like a separate container or entity.

image

The relationship lines would need some testing. Looking at them I feel the inspector looks less readable but maybe they work well in practice.

**Comment by groud** [comment]
OK so at least I think we can ask @djrm to move back class

**Comment by djrm** [comment]
@groud the improvements in this PR are actually obvious, for example now inputs and buttons are actually visible, editing subresources is now managable under the inspector constraints (no need for wide inspector) and several widgets are better organized, the changes here are not perfect, but it is an improvement, also half baked work (even if this is not half baked) gets merged and added all the time (not a justification, but a fact), thats the point of having a dev branch.

|screenshot from 2018-05-24 01-43-42| screenshot from 2018-05-24 01-50-33|

|screenshot from 2018-05-24 01-44-10|screenshot from 2018-05-24 01-50-59|



As i said all the feedback here has been noted, and i plan to do upgrades with the theme update, im not sure how many thing will change with that PR, so i think it will be more work to me do the request changes and maybe redo them when working on updating the theme.

**Comment by groud** [comment]
> Also half baked work (even if this is not half baked) gets merged and added all the time (not a justification, but a fact), thats the point of having a dev branch.

@djrm Sorry, but that's not true. We're developing on the master branch. The development branches are the users' ones. As an example, most of Karrofel's work is on its own fork, and gets merged only when he reaches stepping stones.

Even when we merge WIP, people are asked to solve some problem so that, at least, it is a self-contained improvement. As you know, the issues in the Github are on master, so if you don't solve the problems we pointed out here, we will have (if it's not me other people will) to raise a bunch of issues until it gets fixed someday. This is more work for something we already gave you feedback about here.

I agree most of the changes might need more feedback, but we're at least three of us to think that the class header should be centered as before. For me, this PR should not be merged until, at least, the consensual spotted problems are fixed.

**Comment by groud** [comment]
To moderate my comments, it's my feelings about how we run the project. I still think there are a lot of improvements in this PR. Whether such changes can be accepted with remaining issues or not is more our project manager's decision @akien-mga than mine. ^^

**Comment by reduz** [comment]
Here's my feedback.

1. I really don't like the noodles
2. In this image:
image
  I like the colored X/Y/Z, and I like the boxes appear together, but I think having them horizontal is not good. Vertical was better for Vector2, Vector3, Plane, Quaternion. Even witht that, something here is odd:
image
3. Still unconvinced with curve taking full width. The preset buttons look terrible (my fault I guess), probably will be better to just put a drop down menu with "Preset.." instead, as it was before.
image
4. I would probably make all combos and spins flat (no stylebox) so it's more consistent, this is not so nice:
image
5. Would remove the custom stylebox bg for the subinspectors, so properties can be aligned to the right:
image

**Comment by reduz** [comment]
Another bit of feedback, I think we should probably move the numerical entries to the right. This way, if edited, the entry can appear in the same place as the text (currently not the case) and we still keep the X, Y, Z, etc. indicator.

image

**Comment by reduz** [comment]
for #2. The reasoning for them to be vertical is as follows:
1. The inpector needs to be as compact as possible, and the folding ensures you don't scroll extra. Maybe you work with a large monitor, but a large amount of users use 13" notebooks. On other engines (Unreal, Unity) this battle was lost and using the UIs on a small screens is very unconformable.
2. It's still messy, look at Plane or Quat, they simply don't fit horizontally, and Vector2 is roughly the same space as vertical or horizontal. This only leaves Vector3, which likely works better horizontal, but the sake of consistency, would just put vertical.
3. In a real-life use case, there are only a small amount of properties of this type. Far most of them are single-liners, so it won't really affect your scrolling.

**Comment by groud** [comment]
@reduz. To keep things comfortable on both large and small screens, I think a balanced solution might be to have an adaptable layout: vertical on small width and horizontal on large one. It should not be a lot of work and probably make everyone happy (since from previous comments, it seems like I'm not the only one to find that the horizontal layout looks better).

I'll have look to how it looks on my 13" laptop, and if it is that disturbing.

**Comment by reduz** [comment]
@groud There's nothing much to be assesed honestly. Just look at the screens posted. On subresources (or even main inspector) the horizontal layout suffers and becomes unusable:
image

Look at the room for numbers, there is not any way this is usable.

**Comment by reduz** [comment]
more detail here:
image

**Comment by reduz** [comment]
image

**Comment by reduz** [comment]
Here you can see the values from the screens djrm posted himself, they are all clipped in horizontal.
image

This is why I didnt make those horizontal in the first place.

**Comment by toger5** [comment]
I think one thing that is missing right now, is a common design pattern for each property.

Ill see if I can make an example if it is not obvious what I mean.

**Comment by toger5** [comment]
screen shot 2018-05-24 at 14 46 00
screen shot 2018-05-24 at 14 46 39
There is also a font issue on my machine

**Comment by reduz** [comment]
What I advocate, I believe not only is visually clearer, but also handles sub-resource editing better is the following:
image

**Comment by groud** [comment]
> What I advocate, I believe not only is visually clearer, but also handles sub-resource editing better is the following

I have to admit the vertical layout looks better grouped like that than how I imagined. If it helps on small screens I can handle that ^^

**Comment by Zireael07** [comment]
I like the colored x,y a lot, it will definitely help newcomers.

**Comment by toger5** [comment]
I went just did a mock (vector graphic... no coding) to make clear what I mean by separate the properties more.

untitled 4


- of course color is just temp.
 - I like how the form and sizing of each property turned out.
 - after playing around I also do advocate the "two lines per property" approach. the top line (the white one with the name) can be even thinner In my opinion -> font smaller. The nice thing about a second line is that there is space for actions in the top bar. a got spot for quick help access, copy of the prop name for get() and set() script access, and the requested "reset to last value" button and close button (the close one looks horrible... need a better idea for it...)
 - I think going for the: loose space but make it obvious how much subrecource you are editing is the better approach. Working really deep -> having to spend some time in the inspector -> miking it bigger is smart anyways...
 - No special bg color for the "inheritance" header (node, sprite, control)
### left center right comparison (thoughts)
Basically those three are the last iterations I made during writing this comment... so last one is my fav.

 - It might even be worth to go for a little bit of addition (plus the theme color) tint for deeper layer just to make it easier to have an overview. of course this does not look that good. but it could use the same color than the script tinting and it would at least fit the theme
 - also no border (in the center) looks much better too. 
 - on the right I tried having no section boxes at all when editing a subrecource. This has the advantage that the left side gives you immediate insight on how many levels you are "down" without each lvl creating a subrecourse and a section line. But it makes it less easy to see sections... obviously. So I kept the sections for the most upper lvl.

**Comment by groud** [comment]
> I went just did a mock (vector graphic... no coding) to make clear what I mean by separate the properties more.

That's overkill. Each property is going to take a lot of room for nothing. On your mockup a single color takes the equivalent of 3 lines (or 3 properties) in the current design. That's a no-go for me.

**Comment by toger5** [comment]
another iteration

basically the same design but much more compact.

@reduz I actually tried to adress the issues. Especially that it is hard to have an overview.
And by giving clear elements it is easier to identify what belongs to where.
 - properties are always bright boxes,
 - the bg color (behind properties) is just used to indicate which area you are at. (and not anymore for header backgrounds)
 - And having a clear distinction when a property ends (and the next starts)

screen shot 2018-05-24 at 14 46 00
 
I still don't have an opinion on having horizontally or vertically aligned multi field properties. so I added both. Also for things like textfields and checkboxes there are multiple designs.

But since each property is it's own control now it should be fairly easy to iterate on each of them.

**Comment by groud** [comment]
It's better but I still think the space between each property is not needed.

**Comment by reduz** [comment]
@toger5 space between properties is no needed, and vector2/3/etc needs to be edited vertically, due to the reasons discussed here before your post.

**Comment by toger5** [comment]
yea I agree the vertical position looks better for vector2

**Comment by toger5** [comment]
@groud here is how the gap effects the look:
screen shot 2018-05-24 at 18 49 21
screen shot 2018-05-24 at 18 49 46
screen shot 2018-05-24 at 18 50 55

For me even a little gap makes it so much more obvious what belongs where.
sadly the ordering got messed up...

**Comment by Web-eWorks** [comment]
> @neikeq The clipping in that last screenshot looks weird though. There seems to be space more than enough for another digit...

Godot's LineEdit control has some issues in that regard. It likes to clip the last character, and refuses to scroll it into view.

I agree with @toger5, a 1px gap really helps differentiate between properties.

The colored X/Y/Z labels are nice, but they're just a little too bright. Bring the value down by 10-15% and I think it'd look perfect.

**Comment by djrm** [comment]
@reduz some of the problems can be addressed quite easily like, for example, instead of labels (x,y,z) on small space we could just put a small line with the label color (like in unreal) that would give us 2 more characters on smaller configurations, but really i dont mind much about, it will be a waste for people using wider screens, for example Transforms are 3 columns and have longer label, still, we cant really fix it.
other advantage of having more space for the label is that now keying and revert will look more decent, currently they look ridiculous really, when keying in a small inspector, thw widget is left with no space at all.

about 4 i dont think so, we are mixing up two widgets, the subresource clearly is not a combo box, so i dont see the point in making look like one, that needs to be fixed.

the "noodles" are really bright now, and i figured out a way to make them more discrete, but they are really good to differentiate between the subresources you are editing, we definitely need something like this, specially when editing long subresources, like environment, and materials, it is easy to get lost.

the curve thing, hm i dont really mind much, to me it looks ok, but could be improved with a combo, for presets.

about the alignment, i dont think it will do consistently for int ranges, where you have extra space for the updown, so, im not really sure that is better, as far as i can tell the only alignment that will look the same is centered.

also agree with the category type centered.

just to shed some light in what im planning, i will post some mockups i made

screenshot from 2018-05-24 12-04-14

screenshot from 2018-05-24 12-10-54

**Comment by reduz** [comment]
@djrm ok that looks hell of a lot better. As I mentioned before, the main issue with horizontal xyz is that is that not many decimals can be fit, and this is bothersome.  In your new design it seems you may not need padding, which is a big improvement, so feel free to explore a solution, otherwise you know vertical layout is there waiting for you :P

Plane and Quat will not fit, so those will need a different layout.

**Comment by groud** [comment]
The last design without the padding look like a very good idea. :)
The icon on the top of the header looks weird (an takes a lot of room), I prefer the old one.

Still I feel you're going on the right direction. :+1: 

> Plane and Quat will not fit, so those will need a different layout.

 I agree. For those ones maybe that the old popup might still be possible.

**Comment by tagcup** [comment]
@djrm The labels of the transform shouldn't be x y z; it'd make sense with no labels but it you think labels are better, they should look like

xx xy xz
yx yy yz
zx zy zz

for a matrix representation of rank-2 tensor (which Basis is)

**Comment by tagcup** [comment]
Relevant to the Transform inspector: #10574

**Comment by tagcup** [comment]
The last column of Transform inspector is the translation, I assume, so it would make sense to leave it as x y z

**Comment by djrm** [comment]
@tagcup that is just a mockup, its not correct, it doesnt matter.

**Comment by djrm** [comment]
i cant change them because my editing program has a bug which makes it crash when trying to edit those labels, so i left them that way.

**Comment by volzhs** [comment]
I'm waiting for this. :smile:

**Comment by djrm** [comment]
I wont work on this any longer, sorry, maybe i will try again for the next version,
